### PR TITLE
Update HAVT

### PR DESCRIPTION
**Pull request by shihaocao**
This update actually goes in and implements executing `reset()` and `disable()` based on the cmd_table.

It also will set the read_table based on calling `is_functional()` on all the devices.

In `main.cpp` it didn't display things nicely. New code is marked with a "new code" comment.

Sister FS PR:
https://github.com/pathfinder-for-autonomous-navigation/FlightSoftware/pull/137

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -1,171 +1,198 @@
...
+    registers.ssa.mode = ssa_mode;
+  }
+}
+
+void loop() {
+  update_imu();  DEBUG_printF(",")
+  update_mtr();  DEBUG_printF(",")
+  update_rwa();  DEBUG_printF(",")
+  update_ssa();  DEBUG_printF(",")
+  update_havt(); DEBUG_printF(",") //call update havt last so that reseting devices has a chance to take place before the next control cycle
```
Comment (shihaocao) [inline src/main.cpp]: new code

File: src/main.cpp [hunk 2]
```diff
@@ -1,171 +1,198 @@
...
+  // Wait for sensors to boot up
+  delay(200);
+  // Initialize all subsystems
+  imu::setup();
+  mtr::setup();
+  rwa::setup();
+  ssa::setup();
+}
+
+void update_havt() {
```
Comment (shihaocao) [inline src/main.cpp]: this block is new code, note havt has no setup() block needed above

File: src/main.cpp [hunk 3]
```diff
@@ -1,171 +1,198 @@
...
+//   15 ms till communication can begin
+//   wait 70 ms to take the first read
+//   discard 3 samples at 52 Hz 
+//
+
+#ifdef MAIN_DEBUG
+#define DEBUG
+#endif
+
+#include <adcs_constants.hpp>
```
Comment (shihaocao) [inline src/main.cpp]: new code

File: src/main.cpp [hunk 4]
```diff
@@ -1,171 +1,198 @@
...
+#include <state.hpp>
+#include <state_controller.hpp>
+#include <utl/debug.hpp>
+
+#include <Arduino.h>
+#include <i2c_t3.h>
+#include <lin.hpp>
+
+#include <bitset>
+#include <adcs_havt_devices.hpp>
```
Comment (shihaocao) [inline src/main.cpp]: new code

File: platformio.ini [hunk 5]
```diff
@@ -30,6 +30,20 @@ build_flags =
 src_filter = ${common_env_data.src_filter}
 upload_protocol = ${common_env_data.upload_protocol}
 
+; Builds full flight software for the leader spacecraft.
```
Comment (tanishqaggarwal) [inline platformio.ini]: Consider using the `extends` functionality in the future to reduce copy-paste

File: src/state_controller.cpp [hunk 6]
```diff
@@ -406,8 +406,23 @@ void on_i2c_request() {
...
       //table is stored as an unsigned int, so merely write out to i2c
       endian_write(registers.havt.read_table);
-      std::bitset<32> temp_bitset(registers.havt.read_table);
-      DEBUG_printF("...read havt : ") DEBUG_println(registers.havt.read_table)
+      std::bitset<havt::max_devices> temp_bitset(registers.havt.read_table);
+
+      #ifdef DEBUG
+      char buffer[33];
+      for(int i = 0; i<32; i++){
+        if(temp_bitset.test(31-i))
```
Comment (tanishqaggarwal) [inline src/state_controller.cpp]: Why not use adcs_havt::max_devices here, instead of a hardcoded value?

File: src/main.cpp (hunk 1 above)
Comment (tanishqaggarwal) [reply src/main.cpp]: "Resetting"

File: src/main.cpp [hunk 7]
```diff
@@ -1,171 +1,198 @@
...
+
+  unsigned int command_int;
+  // Attempt atomic copy of the havt command table
+  registers.havt.cmd_flg = CMDFlag::OUTDATED;
+  command_int = registers.havt.cmd_table;
+
+  // Actuate if the copy was atomic
+  if (registers.havt.cmd_flg == CMDFlag::OUTDATED){
+    //I would use the adcs_constants value, but there's a conflicting namespace :L
+    std::bitset<32> temp_command_table(command_int);
```
Comment (tanishqaggarwal) [inline src/main.cpp]: No longer a conflicting namespace issue

File: src/havt.cpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    // command according to cmd_bit
+    if(cmd_bit)
+      dev_ptrs[index]->reset();
+    else{
+      dev_ptrs[index]->disable();
+    }
+  }
+}
+
+void execute_cmd_table(std::bitset<havt::max_devices>& cmd_table){
```
Comment (tanishqaggarwal) [inline src/havt.cpp]: cmd_table can be const

File: platformio.ini (hunk 5 above)
Comment (shihaocao) [reply platformio.ini]: right

**Pull request merged**